### PR TITLE
Add support for running a commandline in another WT window

### LINKED ISSUE #4472: Add support for wt.exe to run commands in an existing Terminal Window
# Description of the new feature/enhancement

I have been playing around with Jetbrains Rider, and when i start a console app with that IDE, it runs the console window inside a docked window in the IDE.

So what i was thinking. How cool would it be, if Windows Terminal allows external applications to start a new Console window in the currently running Windows Terminal instance.

That way whether its VisualStudio or Rider, those IDE's can be configured to start a new Console window inside the Windows Terminal instance.

The way i see this working is: I press F5/ctrl-F5 in visual studio, and instead of opening a new cmd console window, a new tab appears in my running Windows Terminal instance.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

**If you're reading this PR and haven't signed off on #8135, go there first.**

window-management-000

This provides the basic parts of the implementation of #4472. Namely:
* We add support for the `--window,-w <window-id>` argument to `wt.exe`, to allow a commandline to be given to another window.
    * If `window-id` is `0`, run the given commands in _the current window_.
    * If `window-id` is a negative number, run the commands in a _new_ Terminal window.
    * If `window-id` is the ID of an existing window, then run the commandline in that window.
    * If `window-id` is _not_ the ID of an existing window, create a new window. That window will be assigned the ID provided in the commandline. The provided subcommands will be run in that new window.
    * If `window-id` is omitted, then create a new window.

## References
* Spec: #8135
* Megathread: #5000
* Project: projects/5

## PR Checklist
* [x] Closes #4472
* [x] I work here
* [x] Tests added/passed
* [ ] Requires documentation to be updated - **sure does**

## Detailed Description of the Pull Request / Additional comments

Note that `wt -w 1 -d c:\foo cmd.exe` does work, by causing window 1 to change 

There are limitations, and there are plenty of things to work on in the future:
* [ ] We don't support names for windows yet
* [ ] We don't support window glomming by default, or a setting to configure what happens when `-w` is omitted. I thought it best to lay the groundwork first, then come back to that.
* [ ] `-w 0` currently just uses the "last activated" window, not "the current". There's more follow-up work to try and smartly find the actual window we're being called from.
* [ ] Basically anything else that's listed in projects/5.

I'm cutting this PR where it currently is, because this is already a huge PR. I believe the remaining tasks will all be easier to land, once this is in. 

## Validation Steps Performed

I've been creating windows, and closing them, and running cmdlines for a while now. I'm gonna keep doing that while the PR is open, till no bugs remain.

# TODOs
* [x] There are a bunch of `GetID`, `GetPID` calls that aren't try/caught 😬 
  -  [x] `Monarch.cpp`
  -  [x] `Peasant.cpp`
  -  [x] `WindowManager.cpp`
  -  [x] `AppHost.cpp`
* [x] If the monarch gets hung, then _you can't launch any Terminals_ 😨 We should handle this gracefully.
  - Proposed idea: give the Monarch some time to respond to a proposal for a commandline. If there's no response in that timeframe, this window is now a _hermit_, outside of society entirely. It can't be elected Monarch. It can't receive command lines. It has no ID.  
  	- Could we gracefully recover from such a state? maybe, probably not though.
    -  Same deal if a peasant hangs, it could end up hanging the monarch, right? Like if you do `wt -w 2`, and `2` is hung, then does the monarch get hung waiting on the hung peasant?
  - After talking with @miniksa, **we're gonna punt this from the initial implementation**. If people legit hit this in the wild, we'll fix it then.

### CODE REVIEW

File: src/cascadia/Remoting/Monarch.cpp
Comment (skyline75489) [inline src/cascadia/Remoting/Monarch.cpp:101]: Maybe `std::optional<Remoting::IPeasant>` ?

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (skyline75489) [inline src/cascadia/TerminalApp/TerminalPage.cpp:912]: Isn't `std::filesystem::path` a C++17 thing? I didn't know we can use it before.

File: src/cascadia/WindowsTerminal/main.cpp
Comment (skyline75489) [inline src/cascadia/WindowsTerminal/main.cpp:131]: Is the crash by design or just some sort of detect in the framework?

File: src/cascadia/Remoting/Monarch.cpp
Comment (DHowett) [reply src/cascadia/Remoting/Monarch.cpp:101]: Nah. When you have a type that is already nullable, it only confuses things to make it optional as well. Optional adds nullability to non-nullable types. 

You end up in this situation:

```cpp
optional<IUnknown> t;
t = nullopt; // empty
t = null; // NOT EMPTY
t = object; // NOT EMPTY
```

File: src/cascadia/Remoting/Monarch.idl
Comment (skyline75489) [inline src/cascadia/Remoting/Monarch.idl:17]: Darn it. I really want this to be `CreateWindow` 😞

File: src/cascadia/Remoting/WindowManager.cpp
Comment (skyline75489) [inline src/cascadia/Remoting/WindowManager.cpp:209]: I mean, should this be `runningForMonarch`, if the empire is a democratic one. The name `performElection` feels a little off to me.

File: src/cascadia/Remoting/Monarch.cpp
Comment (skyline75489) [reply src/cascadia/Remoting/Monarch.cpp:101]: That's fair. The name `IPeasant` somehow alsoe indicates its nullability.

File: src/cascadia/TerminalApp/AppCommandlineArgs.cpp
Comment (skyline75489) [inline src/cascadia/TerminalApp/AppCommandlineArgs.cpp:193]: Any TODO or issue associated with there?

File: src/cascadia/Remoting/Monarch.cpp
Comment (skyline75489) [inline src/cascadia/Remoting/Monarch.cpp:151]: Maybe revert this logic to reduce indentions.

File: src/cascadia/Remoting/Peasant.cpp
Comment (skyline75489) [inline src/cascadia/Remoting/Peasant.cpp:90]: and -> an

File: src/cascadia/WindowsTerminal/AppHost.cpp
Comment (skyline75489) [inline src/cascadia/WindowsTerminal/AppHost.cpp:549]: Care to find a better naming to avoid `args.Args().Args()` ? I mean, haha. I can't help myself but noticed this 😅

File: src/cascadia/Remoting/WindowManager.cpp
Comment (skyline75489) [inline src/cascadia/Remoting/WindowManager.cpp:39]: Might add some comments about when the thread is *not* joinable.

File: src/cascadia/Remoting/Monarch.cpp
Comment (skyline75489) [inline src/cascadia/Remoting/Monarch.cpp:113]: Why the catch? I see a lot of `GetPID()` that are not surrounded by try-catch.

File: src/cascadia/WindowsTerminal/main.cpp
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/main.cpp:131]: I couldn't tell you. It also might be caused because we're doing weird things with the XamlHost to prevent other crashes, I really can't say.

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:912]: it sure is, and it's _excellent_

File: src/cascadia/TerminalApp/AppCommandlineArgs.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/AppCommandlineArgs.cpp:193]: Oh, you know, I'm gonna yank this dead code. In the scenario where you do `wt -w 100 --maximized` and explicitly want the new window to be created with the ID 100, maximized, then this should work. So I can't have the parser force these args to be exclusive.

File: src/cascadia/Remoting/WindowManager.cpp
Comment (oising) [reply src/cascadia/Remoting/WindowManager.cpp:258]: typo: "iff"

Comment (DHowett) [reply src/cascadia/Remoting/WindowManager.cpp:258]: Ah, but it is [not a typo](https://en.wikipedia.org/wiki/If_and_only_if)!

File: src/cascadia/Remoting/WindowManager.cpp
Comment (carlos-zamora) [inline src/cascadia/Remoting/WindowManager.cpp:384]: This seems to be a very common thing in this PR. We have a lot of loops that look like they could go on forever, but architecturally, they can't. I'm guessing a timeout wouldn't be good here?

**Review by carlos-zamora — CHANGES_REQUESTED** [review]
There's a few files it'd be nice to have an abstract in. Particularly to mention other files to look at too. Like if in "FindTargetWindowArgs" you mentioned what other files use it. Maybe even link the spec too. This would just make reviewing this stuff so much easier.

I got worried about a few potential infinite loops. After reading your PR description, sounds like you're aware of them and gonna fix them later. I'm ok with that. Maybe at least add some kind of logging for now or tag them as a follow-up TODO.

I've had a chance to review #7240. Just need to review #8135. I'll request changes for now to give you a chance to go through my feedback and give me a chance to review that spec. Then we'll regroup and ship this thing. It's awesome!!!

File: src/cascadia/Remoting/WindowManager.cpp
Comment (zadjii-msft) [reply src/cascadia/Remoting/WindowManager.cpp:222]: Yea, we're guaranteed to find one. In the very worst case, _this process_ is also registered to be a monarch, so if _every other process dies in succession_, but this process is still running, then _eventually_, we'll settle on _us_. So if this code is still running, then there's a process that's out there that could be the monarch. That's the fundamental idea with all these loops. I think in practice, you'd have to be closing windows _inhumanly fast_ to iterate more than once.

File: src/cascadia/Remoting/Monarch.cpp
Comment (zadjii-msft) [reply src/cascadia/Remoting/Monarch.cpp:128]: Those should all be fixed now, but go call them out if you see any others

File: src/cascadia/Remoting/Monarch.cpp
Comment (zadjii-msft) [reply src/cascadia/Remoting/Monarch.cpp:171]: Well, a FAIL_FAST is _aggressive_. In the theoretically slim chance that this does occur, then we'd rather not hyper-explode, right? Like, in the chance that you actually hit this, `_peasants.begin()` is going to be the one and only peasant we've talked to, **us**, which is probably the same result that we'd eventually get when we do end up talking to ourself.

If that makes sense

File: src/cascadia/TerminalApp/AppLogic.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/AppLogic.cpp:1191]: Yea, pretty much 🤷‍♂️

File: src/cascadia/Remoting/Monarch.idl
Comment (carlos-zamora) [inline src/cascadia/Remoting/Monarch.idl]: I think you don't actually need to add the constructor to the IDL (and introduce a factory, by extension) if you're only constructing these classes using `winrt::make` and `winrt::make_self`. These constructors are only used if you're trying to construct the objects when you don't have access to the impl type.

File: src/cascadia/Remoting/WindowManager.h
Comment (carlos-zamora) [inline src/cascadia/Remoting/WindowManager.h:6]:
```suggestion
- WindowManager.h
```

**Review by carlos-zamora — APPROVED** [review]
Oh yeah, the constructors and the abstracts make all of this _a lot_ easier to follow. Thanks!

**Comment by zadjii-msft** [comment]
Hey @jsoref, any ideas why the spellcheck bot seems to have exploded in the last couple of days? 147b187 has a comment (that's been minimized) where the bot just marks a ton of misspellings, but on text that's... not in the project?

**Comment by zadjii-msft** [comment]
@jsoref I'm not sure if I've figured it out. #8962 is the first PR who's commit failed the spellcheck. In that commit, some `.PNG` files were added. Not `.png`s, but all-caps `.PNG`s. Looks like the bot might not filter those out? That's at least my theory.

**Comment by zadjii-msft** [comment]
Thanks. Made #9004 to fix that and unblock everyone.

File: src/cascadia/WindowsTerminal/main.cpp
Comment (jsoref) [reply src/cascadia/WindowsTerminal/main.cpp:131]: Fwiw, I believe the generally accepted spelling is `seppuku` in case you want to add it 😉

**Comment by awakecoding** [comment]
@zadjii-msft I would be interested in reparenting Windows Terminal inside another Win32 window, either with the full terminal or just a single tab. We've been doing this successfully in the past in Remote Desktop Manager with powershell.exe and regular console programs, but our generic reparenting code doesn't work with wt.exe. I know this is a separate feature request, but it would be related to the current one as a further improvement. A quick search for "reparenting" didn't yield interesting tickets, would you know if there is a ticket already covering this feature request?

**Comment by zadjii-msft** [comment]
@awakecoding the closest thing to that discussion we've had is over in #7084. That's kinda what you're looking for? If it's not, feel free to open a new issue to track that discussion.

**Comment by jsoref** [comment]
@zadjii-msft: you're going to need to merge w/ main and resolve the file rename (sorry).

Also, did you want someone to fix the case on those `PNG`s?

**Comment by zadjii-msft** [comment]
@jsoref yea I know, thanks. I figured I'd let whoever the second is on this PR give a chance to review it so I only need so switch branches once. Guess I'm just being lazy lol. 

I don't really mind the PNG thing _personally_, I suppose I was mostly just worried that the change from `PNG` to `png` would get Windows all confused, that the files aren't actually different. I guess I didn't know if that mattered with extensions or just filenames. I might be totally misremembering though. If you want to do it, feel free to submit a PR and I'll happily sign off ☺️

**Comment by jsoref** [comment]
I think it's actually dangerous to change just the case of a filename -- so best practice is to pick a different filename (in addition to switching to lowercase): https://stackoverflow.com/questions/1793735/change-case-of-a-file-on-windows/15640313#15640313

File: src/cascadia/Remoting/WindowManager.cpp
Comment (miniksa) [inline src/cascadia/Remoting/WindowManager.cpp:90]: Not only that, the trace logging macros explicitly tell you "DO NOT BE CREATIVE". So this is fine.

File: src/cascadia/Remoting/WindowManager.cpp
Comment (miniksa) [reply src/cascadia/Remoting/WindowManager.cpp:222]: Hm, well, with this explanation, maybe we don't need a timeout on the other one?

**Review by miniksa — APPROVED** [review]
Good enough. There's a lot of moving parts here. I left some brief comments, but I don't think it's anything that can't be solved by adding to your TODO pile. We need SOMETHING to build from to get this done, and this looks like a decent enough base.

File: src/cascadia/Remoting/Monarch.idl
Comment (zadjii-msft) [reply src/cascadia/Remoting/Monarch.idl:17]: That was my guess as well. I aint gonna bother with it, just thought I'd leave the note for someone in 20 years to find

File: src/cascadia/Remoting/Monarch.cpp
Comment (DHowett) [inline src/cascadia/Remoting/Monarch.cpp:82]: the fallback error handler should deal with this, but.. wecan do it ourselves too :)

File: src/cascadia/Remoting/WindowManager.cpp
Comment (DHowett) [reply src/cascadia/Remoting/WindowManager.cpp:41]: (This did not get resolved before merge -- should we have an escape hatch?)

File: src/cascadia/Remoting/WindowManager.cpp
Comment (DHowett) [reply src/cascadia/Remoting/WindowManager.cpp:222]: oh ignore me, i understand now.

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:931]: Will this actually work? if StartingDirectory() is an absolute path or UNC path? what about if it's `""`, the empty string?

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:931]: https://en.cppreference.com/w/cpp/filesystem/path/append

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:931]:
```c++
// On Windows,
path("foo") / "C:/bar";  // the result is "C:/bar" (replaces)
path("foo") / "C:";      // the result is "C:"     (replaces)
path("C:") / "";         // the result is "C:"     (appends, without separator)
path("C:foo") / "/bar";  // yields "C:/bar"        (removes relative path, then appends)
path("C:foo") / "C:bar"; // yields "C:foo/bar"     (appends, omitting p's root-name)
```

**Pull request merged**
